src/HexagonAlignLoads.cpp

```diff
@@ -0,0 +1,217 @@
+#include "HexagonAlignLoads.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "Scope.h"
+#include "Bounds.h"
+#include "ModulusRemainder.h"
+#include "Simplify.h"
+namespace Halide {
+namespace Internal {
+using std::vector;
+
+class HexagonAlignLoads : public IRMutator {
+public:
+    HexagonAlignLoads(Target t, int vs) : target(t), vector_size(vs) {}
+private:
+    Target target;
+    enum AlignCheck { Aligned, Unaligned, NoResult };
+    // size of a vector in bytes.
+    int vector_size;
+    // Alignment info for Int(32) variables in scope.
+    Scope<ModulusRemainder> alignment_info;
+    using IRMutator::visit;
+    ModulusRemainder get_alignment_info(Expr e) {
+        return modulus_remainder(e, alignment_info);
+    }
+    int natural_vector_lanes(Type t) {
+        return vector_size / t.bytes();
+    }
+
+    Expr concat_and_shuffle(Expr vec_a, Expr vec_b, std::vector<Expr> &indices) {
+        Type dbl_t = vec_a.type().with_lanes(vec_a.type().lanes() * 2);
+        Expr dbl_vec = Call::make(dbl_t, Call::concat_vectors, { vec_a, vec_b }, Call::PureIntrinsic);
+        std::vector<Expr> args;
+        args.push_back(dbl_vec);
+        for (int i = 0; i < (int) indices.size(); i++) {
+            args.push_back(indices[i]);
+        }
+        return Call::make(vec_a.type(), Call::shuffle_vector, args, Call::PureIntrinsic);
+    }
+    Expr concat_and_shuffle(Expr vec_a, Expr vec_b, int start, int size) {
+        std::vector<Expr> args;
+        for (int i = start; i < start+size; i++) {
+            args.push_back(i);
+        }
+        return concat_and_shuffle(vec_a, vec_b, args);
+    }
+
+    AlignCheck get_alignment_check(const Ramp *ramp, int host_alignment, int *lanes_off) {
```

Comment:
I found this logic pretty confusing and I can't tell if it's correct. It doesn't seem to take into account the size of the type anywhere. I think a clearer way to do this would be to modify reduce_expr_modulo in ModulusRemainder.cpp to also take a const ```Scope<ModulusRemainder>&```, and then say:

    bool get_alignment_check(Expr base, Type type, int *lanes_off) {
      if (host_alignment % required_alignment) return false;
      int required_alignment_lanes = required_alignment / type.bytes();
      return reduce_expr_modulo(ramp->base, required_alignment_lanes, lanes_off);
    }